### PR TITLE
Fix #8189 - refactor search.c

### LINKED ISSUE #8189: Questionable `search.distance` and missing matches when searching `a+a+b` in `a+a+a+b`
`r2 - -e search.distance=1 -c 'woe 0 @! 9; /x 0908'`
It matches every two bytes. 

The following snippet from `libr/search/search.c` looks questionable to me:
```c
					if (kw->distance < s->distance) {
						kw->idx[kw->distance + 1] = kw->idx[kw->distance];
						kw->distance++;
						hit = true;
					} else {

```

### PR DESCRIPTION
**Pull request by MaskRay**
* fix r_search_mybinparse_update when locating `a+a+b` in `a+a+a+b` where len(a) > 1, e.g. locating `ababc` in `abababc`
* swap the order of the nested loop from

  for (i = 0; i < len; i++)
    r_list_foreach (s->kws, iter, kw)

  to

  r_list_foreach (s->kws, iter, kw)
    for (i = 0; i < len; i++)

  rationale:
  + searching for more needles is rare. the branches used in inner loop harms performance
  + it's cumbersome to bookkeep for each kw in advanced searching algorithms

Fixed tests:

```
[FX]  42  cmd_search: /x with bin mask
[FX]  43  cmd_search: /x with bin mask
```

### CODE REVIEW

**Comment by radare** [comment]
Yeah 3 fixes... but 23 XX

**Comment by MaskRay** [comment]
Reverted this part:

```
diff --git c/libr/core/cmd_search.c i/libr/core/cmd_search.c                                                                                                              
index b8adf2e66..48ea3241d 100644                                                                                                                                         
--- c/libr/core/cmd_search.c                                                                                                                                              
+++ i/libr/core/cmd_search.c                                                                                                                                              
@@ -2010,7 +2010,10 @@ static void do_string_search(RCore *core, struct search_parameters *param) {                                                                       
                                if ((at + bufsz) > param->to) {                                                                                                           
                                        bufsz = param->to - at;                                                                                                           
                                }         
-                               ret = r_io_read_at (core->io, at, buf, bufsz);       
+                               // XXX cause  29  cmd_search: /x wrong entries to fail                                                                                    
+                               // if seek fails we shouldnt read at all             
+                               (void) r_io_seek (core->io, at, R_IO_SEEK_SET);      
+                               ret = r_io_read (core->io, buf, bufsz);  
```

Should be good now but with one less fix.

**Comment by radare** [comment]
Still red

File: libr/search/search.c [hunk 1]
```diff
@@ -427,7 +348,7 @@ R_API RList *r_search_find(RSearch *s, ut64 addr, const ut8 *buf, int len) {
 
 /* --- keywords --- */
 R_API int r_search_kw_add(RSearch *s, RSearchKeyword *kw) {
-	if (!kw) return false;
+	if (!kw->keyword_length) return false;
```
Comment (XVilka) [inline libr/search/search.c]: check both, `kw` can be NULL in theory. Same thing for `s`

File: libr/search/search.c [hunk 2]
```diff
@@ -439,11 +360,11 @@ R_API void r_search_kw_reset(RSearch *s) {
 }
 
 R_API void r_search_reset(RSearch *s, int mode) {
+	R_FREE (s->data);
```
Comment (XVilka) [inline libr/search/search.c]: Check `s` for NULL before dereferencing.

File: libr/search/search.c (hunk 1 above)
Comment (MaskRay) [inline libr/search/search.c]: Unnecessary as call sites have checked this

File: libr/search/search.c (hunk 2 above)
Comment (MaskRay) [inline libr/search/search.c]: This usages is like calling a member function where the first argument denotes the instance. The call site should make this check and it would be too defensive to add such checks.

File: libr/core/cmd_search.c [hunk 3]
```diff
@@ -1924,6 +1924,9 @@ static void do_string_search(RCore *core, struct search_parameters *param) {
 		r_cons_printf ("fs hits\n");
 	}
 	core->search->inverse = param->inverse;
+	if (param->inverse) {
+		core->search->maxhits = 1;
```
Comment (radare) [inline libr/core/cmd_search.c]: Why inverse search forces maxhits to 1?

File: libr/core/cmd_search.c [hunk 4]
```diff
@@ -2349,7 +2353,11 @@ static int cmd_search(void *data, const char *input) {
                 //fin = ini + s->size;
         }
  */
-	maxhits = r_config_get_i (core->config, "search.maxhits");
+	core->search->maxhits = r_config_get_i (core->config, "search.maxhits");
+	// TODO default search.maxhits to UT64_MAX
+	if (!core->search->maxhits) {
```
Comment (radare) [inline libr/core/cmd_search.c]: Wrong. Maxhits=0 is no limit, ut64max is a big limit

File: libr/include/r_search.h [hunk 5]
```diff
@@ -34,8 +34,6 @@ typedef struct r_search_keyword_t {
 	ut8 *bin_binmask;
 	ut32 keyword_length;
 	ut32 binmask_length;
-	ut32 idx[R_SEARCH_DISTANCE_MAX]; // searching purposes
```
Comment (radare) [inline libr/include/r_search.h]: Why do u remove a feature instead of fixing it?

File: libr/search/search.c [hunk 6]
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-						kw->idx[j] = 0;
-						//kw->idx[0] = 0;
-						kw->distance = 0;
-						kw->count++;
-						count++;
-						//s->nhits++;
+	r_list_foreach (s->kws, iter, kw) {
+		ut8 *a = kw->bin_keyword;
+		for (i = 0; i + kw->keyword_length < len; i++) {
+			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
```
Comment (radare) [inline libr/search/search.c]: Strange arithmetics

File: libr/search/search.c [hunk 7]
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
+				for (j = 1; j < kw->keyword_length &&
```
Comment (radare) [inline libr/search/search.c]: Ugly for

File: libr/search/search.c [hunk 8]
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-	int i, kw_len;
+	int i, j, m = kw->keyword_length - 1, count = 0;
 	ut8 ch;
 
-	kw_len = kw->keyword_length - 1;
-
-	if (kw_len < 0)
-		return false;
-
-	for (i = 0; i < 256; i++)
```
Comment (radare) [inline libr/search/search.c]: Missing braces

File: libr/search/search.c [hunk 9]
```diff
@@ -125,258 +131,172 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
+					i += kw->keyword_length;
+					goto next;
+				}
 			}
 		}
-		ch = buf[pos + kw_len];
-		pos += bad_char_shift[kw->icase?tolower(ch):ch];
+		ch = buf[i + m];
+		i += bad_char_shift[kw->icase ? tolower (ch) : ch];
+	next:;
```
Comment (radare) [inline libr/search/search.c]: Labels doesnt require ;

File: libr/core/cmd_search.c (hunk 3 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: This is exactly what the code did before.

File: libr/core/cmd_search.c (hunk 4 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: `ut64max` can be seen as infinity and the limit cannot be reached. This simplifies code otherwhere and there would be many more other issues before the limit was reached.

File: libr/search/search.c (hunk 2 above)
Comment (XVilka) [inline libr/search/search.c]: This is not C++.

File: libr/include/r_search.h (hunk 5 above)
Comment (MaskRay) [inline libr/include/r_search.h]: It was implemented in such a wrong way that I could not guess its intention.

File: libr/search/search.c (hunk 6 above)
Comment (MaskRay) [inline libr/search/search.c]: `/d` was incorrect if `/d 0101` was intended to find `01 02 03`. This is corrected now.

File: libr/search/search.c (hunk 7 above)
Comment (MaskRay) [inline libr/search/search.c]: Move `j++` inside of `{}`?

File: libr/search/search.c (hunk 9 above)
Comment (MaskRay) [inline libr/search/search.c]: The last statement in a scope requires `;`.

File: libr/search/search.c (hunk 8 above)
Comment (MaskRay) [inline libr/search/search.c]: Fixed but not yet pushed to wait for travis.

File: libr/search/search.c (hunk 7 above)
Comment (MaskRay) [inline libr/search/search.c]: Fixed but not yet pushed to wait for travis.

File: libr/search/search.c (hunk 2 above)
Comment (MaskRay) [inline libr/search/search.c]: The check was not there before my change. Do that in another commit if you really care.

File: libr/include/r_search.h (hunk 5 above)
Comment (MaskRay) [inline libr/include/r_search.h]: Brought back.

File: libr/core/cmd_search.c (hunk 3 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: This behavior was implemented in `R_API int r_search_mybinparse_update(void *_s, ut64 from, const ut8 *buf, int len) {` and there are tests testing this. I just moved it outside .

File: libr/search/search.c (hunk 8 above)
Comment (MaskRay) [inline libr/search/search.c]: Pushed

File: libr/core/cmd_search.c (hunk 4 above)
Comment (radare) [inline libr/core/cmd_search.c]: imho having 0 maxhits is more logic and easy to initialize than a value that is not zero. and it was already like this. also by the user it is easier to type e search.maxhits=0 than e search.maxhits=0xffffffffffffffffff

File: libr/include/r_search.h (hunk 5 above)
Comment (radare) [inline libr/include/r_search.h]: we can discuss this thing in a separate issue/pr, better change only 1 thing on each PR, this way it's easier to review

File: libr/search/search.c [hunk 10]
```diff
@@ -42,6 +50,7 @@ R_API RSearch *r_search_new(int mode) {
 R_API RSearch *r_search_free(RSearch *s) {
 	if (!s) return NULL;
```
Comment (radare) [inline libr/search/search.c]: missing braces

File: libr/search/search.c (hunk 6 above)
Comment (radare) [inline libr/search/search.c]: can you add tests for this or unmark the BROKEN in r2r? ill merge that before merging this.

thanks

File: libr/search/search.c (hunk 9 above)
Comment (radare) [inline libr/search/search.c]: it will be probably easier to read if you put this label at the begining of the loop. also we may assume i>0 otherwise this can result in an infinite loop.

File: libr/search/search.c [hunk 11]
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-			offset += match_pos + kw->keyword_length;
-			kw->count++;
-			count++;
-			/* Stop at the first occurrence */
-			if (s->inverse) {
-				return -1;
-			}
+		longest = R_MAX (longest, kw->keyword_length);
+	}
+	if (!longest) return 0;
```
Comment (radare) [inline libr/search/search.c]: missing braces

File: libr/search/search.c [hunk 12]
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
+	if (s->data) {
+		left = s->data;
+		if (left->end != from) {
+			left->len = 0;
 		}
+	} else {
+		left = malloc (sizeof(RSearchLeftover) + 2 * (longest - 1));
+		if (!left) return -1;
```
Comment (radare) [inline libr/search/search.c]: missing braces

File: libr/search/search.c [hunk 13]
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-							}
-						}
-						//s->nhits++;
-					}
+		for (i = 0; i + kw->keyword_length <= len; i++) {
+			if (brute_force_match(s, kw, buf, i) != s->inverse) {
+				if (!r_search_hit_new (s, kw, from + i)) return -1;
+				kw->count++;
+				count++;
+				if (++s->nhits >= s->maxhits) return count;
```
Comment (radare) [inline libr/search/search.c]: missing braces

File: libr/search/search.c [hunk 14]
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-						if (s->overlap) {
-							if (kw->keyword_length > 1) {
-								i -= kw->keyword_length - 1;
-							}
-						}
-						//s->nhits++;
-					}
+		for (i = 0; i + kw->keyword_length <= len; i++) {
+			if (brute_force_match(s, kw, buf, i) != s->inverse) {
+				if (!r_search_hit_new (s, kw, from + i)) return -1;
```
Comment (radare) [inline libr/search/search.c]: missing braces

File: libr/search/search.c [hunk 15]
```diff
@@ -125,250 +132,203 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-						count++;
-						if (s->overlap) {
-							if (kw->keyword_length > 1) {
-								i -= kw->keyword_length - 1;
-							}
-						}
-						//s->nhits++;
-					}
+		for (i = 0; i + kw->keyword_length <= len; i++) {
+			if (brute_force_match(s, kw, buf, i) != s->inverse) {
```
Comment (radare) [inline libr/search/search.c]: missing space before (

File: libr/search/search.c (hunk 1 above)
Comment (radare) [inline libr/search/search.c]: missing braces

Comment (MaskRay) [inline libr/search/search.c]: IMHO using curly braces everywhere does not enhance readability, especially for short conditional statements such as:
```c
if (!kw->keyword_length) {
  return false;
}
```

Linux coding style allows:
```c
        if (!buffer)
                return -ENOMEM;
```

Google C++ Style Guide (https://google.github.io/styleguide/cppguide.html#Conditionals) allows the statement put on the same line if it enhances readability:

```c
if (x == kFoo) return new Foo();
```

File: libr/core/cmd_search.c (hunk 4 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: If `search.maxhits` is zero (default), `core->search->maxhits` will be `UT64MAX`.

File: libr/include/r_search.h (hunk 5 above)
Comment (MaskRay) [inline libr/include/r_search.h]: `idx[R_SEARCH_DISTANCE_MAX]` gets in the way when I made the PR. It introduced another nested loop which I did not know how to put it back in the code after the refactoring as the loop order had been swapped.

File: libr/search/search.c (hunk 9 above)
Comment (MaskRay) [inline libr/search/search.c]: Moved `next` to the beginning of the loop.

`i>=0` is a loop invariant.

File: libr/search/search.c (hunk 10 above)
Comment (MaskRay) [inline libr/search/search.c]: IMHO the following two alternatives enhance readability. Also see my comment above.
```c
if (!s) return NULL;
```

```c
if (!s)
  return NULL;
```

File: libr/search/search.c (hunk 15 above)
Comment (MaskRay) [inline libr/search/search.c]: Fixed to be consistent with other code, though I always think space between function name and `(args` is bizarre...

**Comment by MaskRay** [comment]
Resolved for most style issues.

`if (!s) return NULL;`

Though mandatory curly braces around simple conditional statements are still unacceptable to me.
```c
if (!s) {
  return NULL;
}
`

File: libr/search/search.c (hunk 1 above)
Comment (radare) [inline libr/search/search.c]: Not the place to discuss this and it was already discussed several times. You are late to the party and we spent 2 years cooking a coding style that fits well for most of us. So just follow it

File: libr/search/search.c (hunk 10 above)
Comment (radare) [inline libr/search/search.c]: See coding style and be consistent. This has been discussed zillions of times and im not gonna explain the reasons again

**Comment by MaskRay** [comment]
Rebased on HEAD and added braces.

**Comment by radare** [comment]
The braces in oneliner blocks they look bad indeed. But they clearly show something that is important and eases the reading, also exposes something that could be expressed in a better way by following functional programming or raii techniques. Also gotos are not considered a good practicw becayse makes the reader jump up and down randomly. But they sometimes help on reducing code that involves error checking.

Ive been coding for more than 20 years and those are some insights i've learned. I would never do those things 15 years ago. So consider it an advice for the future and for consistency. Perl for example imposes braces on all blocks. Even one liners and the reasons vary, but one of them is readability.

Your future yourself will be pleased to follow some consistent coding styles. Trust me ;)

File: libr/search/search.c [hunk 16]
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-	ut8 a, b;
-
-	kwidx = kw->idx[kwidx];
-	k = (kwidx % kw->binmask_length);
-	a = *buf;
-	b = kw->bin_keyword[kwidx];
-	if (kw->icase) {
-		a = tolower (a);
-		b = tolower (b);
+static inline bool brute_force_match(RSearch *s, RSearchKeyword *kw, const ut8 *buf, int i) {
```
Comment (radare) [inline libr/search/search.c]: Inlining a 100 line function? No thanks

File: libr/core/cmd_search.c [hunk 17]
```diff
@@ -2352,7 +2356,10 @@ static int cmd_search(void *data, const char *input) {
                 //fin = ini + s->size;
         }
  */
-	maxhits = r_config_get_i (core->config, "search.maxhits");
+	core->search->maxhits = r_config_get_i (core->config, "search.maxhits");
+	if (!core->search->maxhits) {
```
Comment (radare) [inline libr/core/cmd_search.c]: Remove this

File: libr/core/cmd_search.c (hunk 3 above)
Comment (radare) [inline libr/core/cmd_search.c]: Just keep the old behaviour. Theres no sense in searching for 0 keywords and ut64max is such a bad number to type for a human to specify that theres no limit

File: libr/core/cmd_search.c [hunk 18]
```diff
@@ -2716,7 +2723,7 @@ static int cmd_search(void *data, const char *input) {
 		if (input[0] == 'j') {
 			json = true;
 		}
-	/* pass-thru */
+		// fallthrough
```
Comment (radare) [inline libr/core/cmd_search.c]: The reason for that comment is to avoid an error in scan.coverity not sure if that change will be also accepted by them but we can try

File: libr/search/search.c [hunk 19]
```diff
@@ -28,6 +35,7 @@ R_API RSearch *r_search_new(int mode) {
 	s->string_max = 255;
 	s->string_min = 3;
 	s->hits = r_list_new ();
+	s->maxhits = UT64_MAX;
```
Comment (radare) [inline libr/search/search.c]: Should be =0

File: libr/search/search.c [hunk 20]
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
+	if (!longest) {
+		return 0;
+	}
+	if (s->data) {
+		left = s->data;
+		if (left->end != from) {
+			left->len = 0;
+		}
+	} else {
+		left = malloc (sizeof(RSearchLeftover) + 2 * (longest - 1));
```
Comment (radare) [inline libr/search/search.c]: Missing space and maybe use calloc instead to avoid uninit buffer

Comment (radare) [inline libr/search/search.c]: Longest can be <2? If so that code can bring some problems

Comment (radare) [inline libr/search/search.c]: Better use calloc instead or * to avoid integer overflows

File: libr/search/search.c [hunk 21]
```diff
@@ -125,250 +134,217 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
-					break;
+		i = !s->overlap && from - kw->last < left->len ? kw->last + left->len - from : 0;
+		for (; i + kw->keyword_length <= len1 && i < left->len; i++) {
+			if (brute_force_match (s, kw, left->data, i) != s->inverse) {
+				if (!r_search_hit_new (s, kw, from + i - left->len)) {
+					return -1;
+				}
+				kw->count++;
+				count++;
+				if (++s->nhits >= s->maxhits) {
```
Comment (radare) [inline libr/search/search.c]: Im not a big fan of preincrements inside if conditionals better put yhat in the line above

File: libr/search/search.c [hunk 22]
```diff
@@ -427,7 +406,9 @@ R_API RList *r_search_find(RSearch *s, ut64 addr, const ut8 *buf, int len) {
 
 /* --- keywords --- */
 R_API int r_search_kw_add(RSearch *s, RSearchKeyword *kw) {
-	if (!kw) return false;
```
Comment (radare) [inline libr/search/search.c]: Null deref regression

File: libr/search/search.c (hunk 2 above)
Comment (radare) [inline libr/search/search.c]: Add the check

File: libr/search/search.c [hunk 23]
```diff
@@ -439,11 +420,11 @@ R_API void r_search_kw_reset(RSearch *s) {
 }
 
 R_API void r_search_reset(RSearch *s, int mode) {
+	R_FREE (s->data);
 	r_list_purge (s->hits);
-	s->nhits = 0;
-	s->hits = r_list_new ();
+	s->hits = r_list_newf (free);
 	if (!s->hits) return;
```
Comment (radare) [inline libr/search/search.c]: Missing braces

File: libr/search/search.c [hunk 24]
```diff
@@ -439,11 +420,11 @@ R_API void r_search_kw_reset(RSearch *s) {
...
-	s->hits->free = free;
+	s->nhits = 0;
 	r_search_kw_reset (s);
 	if (!r_search_set_mode (s, mode))
```
Comment (radare) [inline libr/search/search.c]: Missing braces

File: libr/core/cmd_search.c (hunk 3 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: The old behavior is stopping at the first occurrence:

```c
-                       /* Stop at the first occurrence */
-                       if (s->inverse) {
-                               return -1;
-                       }
```

Then `r_search_update` returns `-1` and `do_string_search` in cmd_search.c breaks the loop.

File: libr/search/search.c (hunk 6 above)
Comment (MaskRay) [inline libr/search/search.c]: https://github.com/radare/radare2-regressions/pull/959

I do not see how to make `/d` output one more byte for each hit.

File: libr/search/search.c (hunk 16 above)
Comment (MaskRay) [inline libr/search/search.c]: Done

File: libr/search/search.c (hunk 20 above)
Comment (MaskRay) [inline libr/search/search.c]: Its `sizeof(RSearchLeftover) + 2 * (longest - 1)`, more disasters should have happened before the integer overflow kicks in.

File: libr/search/search.c (hunk 21 above)
Comment (MaskRay) [inline libr/search/search.c]: Done

File: libr/search/search.c (hunk 22 above)
Comment (MaskRay) [inline libr/search/search.c]: Now `	if (!kw || !kw->keyword_length) {`

I believe the call site should do the check, not here.

File: libr/search/search.c (hunk 24 above)
Comment (MaskRay) [inline libr/search/search.c]: Done

File: libr/core/cmd_search.c (hunk 17 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: Removed

File: libr/core/cmd_search.c (hunk 18 above)
Comment (MaskRay) [inline libr/core/cmd_search.c]: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html -Wimplicit-fallthrough=n

`fallthrough` is recognized but not `pass-thru`.

File: libr/search/search.c (hunk 23 above)
Comment (MaskRay) [inline libr/search/search.c]: Done ( though i hadn't touched it)

File: libr/core/cmd_search.c (hunk 18 above)
Comment (radare) [inline libr/core/cmd_search.c]: oki :)

File: libr/search/search.c [hunk 25]
```diff
@@ -9,6 +9,12 @@
 
 R_LIB_VERSION (r_search);
 
+typedef struct {
+	ut64 end;
+	int len;
+	ut8 data[];
```
Comment (radare) [inline libr/search/search.c]: use ut8 *data

Comment (MaskRay) [inline libr/search/search.c]: `ut8 *data` (avoiding flexible array member) leads to one extra `malloc`, more error check with no benefit.

Comment (MaskRay) [inline libr/search/search.c]: It is also supported by tcc.

Comment (condret) [inline libr/search/search.c]: use * instead of [] , it is more portable, especially for older compilers. Use [] only for fixed sizes

Comment (MaskRay) [inline libr/search/search.c]: I'd like to accept a feature if compiles on gcc, clang and MSVC 2015.
Without  flexible array member, you'll do two `malloc`, the second for `ut8 *data`, and if it returned `NULL`, you had to clean the struct and return `NULL`.

Comment (MaskRay) [inline libr/search/search.c]: Yes, I'm also talking to `r_strbuf`

```c
R_API bool r_strbuf_set(RStrBuf *sb, const char *s) {
	int l;
	if (!sb) {
		return false;
	}
	if (!s) {
		r_strbuf_init (sb);
		return true;
	}
	l = strlen (s);
	if (l >= sizeof (sb->buf)) {
		char *ptr = sb->ptr;
		if (!ptr || l+1 > sb->ptrlen) {
			ptr = malloc (l + 1);
			if (!ptr) {
				return false;
			}
			sb->ptrlen = l + 1;
			sb->ptr = ptr;
		}
		memcpy (ptr, s, l+1);
	} else {
		sb->ptr = NULL;
		memcpy (sb->buf, s, l+1);
	}
	sb->len = l;
	return true;
}
```

**Comment by MaskRay** [comment]
```c
typedef struct {
	int len;
	char *ptr;
	int ptrlen;
	char buf[64];
} RStrBuf;

R_API bool r_strbuf_set(RStrBuf *sb, const char *s) {
	int l;
	if (!sb) {
		return false;
	}
	if (!s) {
		r_strbuf_init (sb);
		return true;
	}
	l = strlen (s);
	if (l >= sizeof (sb->buf)) {
		char *ptr = sb->ptr;
		if (!ptr || l+1 > sb->ptrlen) {
			ptr = malloc (l + 1);
			if (!ptr) {
				return false;
			}
			sb->ptrlen = l + 1;
			sb->ptr = ptr;
		}
		memcpy (ptr, s, l+1);
	} else {
		sb->ptr = NULL;
		memcpy (sb->buf, s, l+1);
	}
	sb->len = l;
	return true;
}
```

This can be avoided if flexible array member is used.

**Comment by radare** [comment]
you cant resize a flexible array member, you have to resize the whole pointer, so ends up breaking the api because you would need to pass RStrBuf** or return the new RStrBufer delaying the responsability to control the new pointer to the caller. Seems like msvc is also supporting this, and this will reduce the memory usage, but breaking api is probably the main concern for me

File: libr/search/search.c (hunk 20 above)
Comment (radare) [inline libr/search/search.c]: its not about the worst case scenario, its about good practices

File: libr/search/search.c (hunk 22 above)
Comment (radare) [inline libr/search/search.c]: an if doesnst hurts and prevents anoying null derefs, compilers can always optimize that stuff with lto

File: libr/search/search.c [hunk 26]
```diff
@@ -127,250 +136,220 @@ R_API int r_search_hit_new(RSearch *s, RSearchKeyword *kw, ut64 addr) {
...
 
+// Supported search variants: overlap
 R_API int r_search_deltakey_update(void *_s, ut64 from, const ut8 *buf, int len) {
 	RListIter *iter;
-	unsigned char pch = 0;
 	int i, j, count = 0;
 	RSearch *s = (RSearch*)_s;
+	RSearchKeyword *kw;
 
-	for (i=0; i<len; i++) {
```
Comment (radare) [inline libr/search/search.c]: so you are only checking  the first char of the buffer, not all of them?

Comment (MaskRay) [inline libr/search/search.c]: The code does check all of them.

```c
		for (i = 0; i + kw->keyword_length < len; i++) {
			if ((ut8)(buf[i+1] - buf[i]) == a[0]) {
				j = 1;
				while (j < kw->keyword_length && (ut8)(buf[i+j+1] - buf[i+j]) == a[j]) {
					j++;
				}
				if (j == kw->keyword_length) { ////////////// all of them

```

File: libr/search/search.c (hunk 22 above)
Comment (MaskRay) [inline libr/search/search.c]: The call sites should do the check.

```c
	case 'a': if (input[1]) {
...........
				r_search_kw_add (core->search,
					r_search_keyword_new_hexmask (kwd, NULL)); ///// it should check if new_hexmask returns NULL
				r_search_begin (core->search);
				free (kwd);
				dosearch = true;
```

The call site are closer to the user and have more knowledge about the interactive context. It can give useful warnings/errors and it should have do that. Then such check in the search.c side will be redundant. If the library is full of such `if (!xxx) return;` checks, there will not be much integrity guarantee and you'll tend to distrust everything.

Comment (MaskRay) [inline libr/search/search.c]: I was really, really sad when I saw the following code for the first time:

```c
#define r_list_foreach(list, it, pos)\
	if (list)\
		for (it = list->head; it && (pos = it->data); it = it->n)
```

`pos = it->data` is dangerous, what if one of the elements is `NULL`? It will stop iterating at that element, ignoring the rest.

`if (list)` will you miss something if the for loop has not been executed? The (no-null-dereference) illusion comes with the cost that more trouble may accompany.

File: libr/search/search.c (hunk 20 above)
Comment (MaskRay) [inline libr/search/search.c]: Changed to `		left = malloc (sizeof(RSearchLeftover) + (size_t)2 * (longest - 1));`

It will never cause trouble where `size_t` is 64-bit and `longest` is 32-bit.

**Comment by radare** [comment]
once master gets green again i would like this pr to be rebased and i'l have a look again, sorry for the lag, crazy days irl

**Comment by MaskRay** [comment]
It makes things simpler if this PR can be decided "to be or not to be" earlier. There are also new challenges when siol lands. Stacking maps need different treatment other than the current approach.

**Comment by MaskRay** [comment]
Rebased on master.

@alvarofe Your recent `bool r_io_read_at` stuff makes search more broken.

**Comment by condret** [comment]
@MaskRay use the al_read here, this is the way to go

**Comment by MaskRay** [comment]
Rebased.

A
G
A
I
N

File: libr/core/cmd_search.c [hunk 27]
```diff
@@ -1854,6 +1854,10 @@ static void do_string_search(RCore *core, struct search_parameters *param) {
 		r_cons_printf ("fs hits\n");
 	}
 	core->search->inverse = param->inverse;
+	// TODO Bad but is to be compatible with the legacy behavior
+	if (param->inverse) {
+		core->search->maxhits = 1;
```
Comment (radare) [inline libr/core/cmd_search.c]: Why

File: libr/search/search.c (hunk 22 above)
Comment (radare) [reply libr/search/search.c]: Data should never be null.

**Comment by radare** [comment]
im merging, but im not happy with removing features or hardcoding maxhits, but this thing got too far. please FOCUS, we need to fix all the issues related to the siol merge, and without having travis green its a pain to keep merging more stuff that cant be properly checked without that. several more tests needs to be updated after my changes in search.in to make it correct.

**Comment by MaskRay** [comment]
I was not removing features or harcoding maxhits (it existed before the refactor)

`maxhits` was used as a decreasing counter. I made it constant in the search loop and added another field `nhits` to record the number of hits.
I was not removing features. The so-called "search purpose" was used to implement `search.distance`, but it was incorrect and harmed the performance drastically (O(|haystack|*|needle|*|search.distance|), with some coined test cases, it fell into an infinite loop). So I had to remove it and replaced it with a seemingly-correct leftover.

**Comment by radare** [comment]
Ok
